### ISSUE TITLE
(RFC) Rewrite index.html in Elm

### ISSUE THREAD

**Issue by JustusAdam** [elm-lang/elm-reactor#121]
As listed in [projects](https://github.com/elm-lang/projects) the index page requires a rewrite in Elm.

My proposition would be to move to an asynchronous javascript approach where the actual page would be served only once and navigating the directory tree would be ajax requests to the server, reloading (parts of) the page.

I'd like to try and write that, however that requires changes to the server code, which is why I'd like to collect some feedback first if that is even a sensible approach.

**Comment by evancz** [comment]
I was thinking we'd have an Elm file that takes in some JSON and displays it (exactly how package.elm-lang.org works) so there's just one file describing how to show a navigation page, but JSON is generated based on directory structure.

I had imagined sending that Elm file and some JSON for each directory. I guess it's conceivable to do it all with JSON after the initial load, but my intuition is to split it into steps. I'd wager that going for the Elm + JSON for every page and *then* trying to do it as a single page thing is more likely to be successful.

Also, does a single page have significant benefits for us? I'd want to be clear about what we are gaining before going down that route!

**Comment by JustusAdam** [comment]
I was honestly just thinking about performance. Resending the same file over and over again seems quite redundant.

However, if the backend just returns 304 and the browser uses the same elm file and just a new json there shouldn't be much difference.
One of the reasons I suggested it is because I've written an asynchronous page like that before (in elm) and sort of just defaulted to the concept.

So we're basically just talking about a rewrite in elm (and elm-html, because it's awesome)?

That could be arranged I think.

How'd you propose the JSON is added? I'd like to not compile it into the elm code, so that we get fewer recompiles and less load. Where would be the best place to put it? Perhaps just send a separate `_data.js` file that only contains the json data and exposing it via port?

**Comment by evancz** [comment]
> I was honestly just thinking about performance.

Since the server is running locally, I don't think there is a human that exists that'd notice the difference :)

> So we're basically just talking about a rewrite in elm (and elm-html, because it's awesome)?

That's my priority!

> Perhaps just send a separate `_data.js` file that only contains the json data and exposing it via port?

Yeah, this is roughly what package.elm-lang.org does. Check out view-source:http://package.elm-lang.org/packages/elm-lang/core/2.1.0/ 

It is generating this little JS snippet depending on the request sent by the user:

```javascript
var context = { user: "elm-lang", name: "core", version: "2.1.0", versionList: ["2.1.0","2.0.1","2.0.0","1.1.1","1.1.0","1.0.0"] };
var page = Elm.fullscreen(Elm.Page.Package, { context: context });
```

And the `Page.Package` Elm module is able to figure everything else out based on that information.

Again, especially with `elm-reactor` speed is not a big factor for the foreseeable future. The rate of queries is limited to what a single human can do, and the delay just has to be lower than what humans can perceive. Things would have to go very very wrong for us to get into trouble here!

**Comment by JustusAdam** [comment]
I'll make a prototype. I'll probably delay integrating it with the server for now, until the prototype works as expected.

**Comment by JustusAdam** [comment]
Okay. A very first prototype is done [here](https://github.com/JustusAdam/elm-reactor-index).

It focuses mostly on making a page only depending on a `_data.js` file.

**Comment by vilterp** [comment]
Hey @JustusAdam, thanks for working on this! (I'm working on the reactor this summer with Evan)

A couple thoughts:

1. So we can build and evaluate your code, could you elaborate in the readme on how to install Compass and use to build the stylesheets? A Makefile would be great.
2. In the longer term though, It would be nice if we could do the styles without Compass (i.e. in plain CSS or inline on elements with pure Elm), because (a) Compass adds a Ruby build dependency to Reactor, which makes building it more complicated, and (b) it just seems simpler to use Elm to do whatever style composition we need (w/ inline styles), since Compass's variables and other constructs seem to be duplicating power that's already in Elm.

**Comment by JustusAdam** [comment]
Interesting thought ... 

I've never tried elm's inline styles. I'll have a look at it.

My original plan was when the page gets put in the reactor I do a compass build and add the css, not the sass files.

**Comment by JustusAdam** [comment]
Sass was just meant to help me develop the style faster and more reliable.

**Comment by vilterp** [comment]
That's totally fine, just some help figuring out how to build it for now would be appreciated so we can check out what you made. Thanks again!

**Comment by JustusAdam** [comment]
Added some more install documentation with #ef376cd6

**Comment by vilterp** [comment]
![screen shot 2015-07-16 at 1 07 04 pm](https://cloud.githubusercontent.com/assets/7341/8734036/23e2bb56-2bbc-11e5-9eb0-d649784ee68f.png)
 Looks good! Uploading this screenshot for convenience

**Comment by JustusAdam** [comment]
Yes. I actually wanted to add one to the readme, but couldn't be bothered to figure out how  ... :laughing:

**Comment by JustusAdam** [comment]
I did some more work on the file display and added a current screenshot to the [repository](https://github.com/JustusAdam/elm-reactor-index) so you can easily see what's going on.

**Comment by evancz** [comment]
This looks great @JustusAdam! Your screenshot looks really amazing :D

I also share the concern about adding a dependency for the CSS stuff. What steps are left besides that for getting this ready for a PR to elm-reactor? Is it close?

**Comment by evancz** [comment]
I think my one visual concern would be that there are so many different icons that it becomes confusing. Looking at the normal github icons, there's just a generic picture for files.

![screen shot 2015-07-17 at 1 28 33 pm](https://cloud.githubusercontent.com/assets/1658058/8756548/e864eb4a-2c87-11e5-932b-294131547c18.png)

It feels easier for me to find the files I want in that one. I'm not sure if that's just familiarity, but my intuition is that the icons are visually distracting, but diverse enough that they don't help me pick things out. Whereas the simple division between blue folders and white files helps me quickly jump to interesting stuff.

**Comment by JustusAdam** [comment]
I thought it would be a nice idea, to be able to gather more quickly the file you re looking for. So if you want config, you'd look for the wrench and if you want code you'd look for the </> symbol.

Here are two slightly more realistic examples of how It'd look. The first screenshot was very much designed to show off as many as possible.

![elm-reactor-index-2](https://cloud.githubusercontent.com/assets/5824532/8756771/eac33f7a-2cd4-11e5-852c-7ebd273d755e.png)

![elm-reactor-index-3](https://cloud.githubusercontent.com/assets/5824532/8756774/f0bbec10-2cd4-11e5-8018-405929ccf4c1.png)

I don't know when we can merge, because the server will have to be altered also. This was only the Frontend. I'll have to look at reactors backend again and change the server so that it provides the right json data for the site.

I did my best to decouple main page and data so that shouldn't actually be too hard. Depends on the reactor backhand though.

As for the css: Technically neither sass nor elm are actually required for the page. Both can be statically compiled and don't have to be changed after that. When the page is served neither sass nor elm needs compling, since I moved the actual data out of the elm code. If you'd like just the css in the reactor that's fine and the elm-code can either be dropped in just as javascript or compiled on build or whatever seems best.

**Comment by JustusAdam** [comment]
Ah. I finally found that having a record field called 'package' crashed the runtime. Now my package information view actually works.

![elm-reactor-index-2](https://cloud.githubusercontent.com/assets/5824532/8757628/1e964d36-2cdc-11e5-81f9-f7bc4220c391.png)

**Comment by evancz** [comment]
This looks so gosh darn good!

I've heard of someone else saying the name `package` was causing trouble, but without an issue coming out of it. If I write programs like [this](http://pastebin.com/6HtyQxvZ), I don't see the error. If you can get it down to a small example, please let us know about it on elm-compiler (or maybe elm-lang/core if that's the culprit)

**Comment by JustusAdam** [comment]
I found some minimal code to produce the bug and [reported it](https://github.com/elm-lang/elm-compiler/issues/994).

**Comment by ohanhi** [comment]
Hi @JustusAdam, this is looking great! I was just about to start working on the same thing, but then found that you're (more than) halfway there already. So I forked your repo and will try to inline the styles you've created into Elm.

**Comment by JustusAdam** [comment]
Nice. Cool. Then I'll have a look at the backend server.

**Comment by ohanhi** [comment]
Okay, I made a pull request containing all the styles from Sass now inline within the Elm code. There seemed to be a conflict with your changes, but it looked like quite an easy resolve.

**Comment by JustusAdam** [comment]
Very nice job on inlining the style. 

The conflict btw comes from me making changes to the style before, because on folders with few files the dependencies would get out of place. 

I merged the changes back in and also removed all classes, since they really aren't necessary anymore.

**Comment by JustusAdam** [comment]
The backend server is almost ready, it serves the page just fine, but I still need to figure out how I'll make it serve the assets. @evancz and @vilterp I have an Issue open on the reactor on serving assets #128 that I'd really like to get just a little bit of feedback back on, before I hardcode the icons in the server.

One thing I haven't quite solved (mostly due to lack of time) is how to serve/integrate the compiled elm. Is there a way to compile it once when the server is being built and then just serve the js? 

Because that'd be ideal. I'll have a look at the other Elm code in the reactor and see how it's done there.

**Comment by evancz** [comment]
I responded on #128. I'd go with serving it under `_reactor/` and we can make decisions about the "right" way to do that separately from this project. At that point it'll be a pretty easy change.

The setup code lives [here](https://github.com/elm-lang/elm-reactor/blob/master/Setup.hs) and it builds and bundles everything at compile time. So your stuff should get brought in there.

**Comment by JustusAdam** [comment]
It is done #129

**Comment by JustusAdam** [comment]
Was merged.
